source/AI.cpp

```diff
@@ -4724,3 +4738,49 @@ void AI::UpdateOrders(const Ship &ship)
 		order.targetSystem = ship.GetSystem();
 	}
 }
+
+// Job / Mission NPC blocks may use keywords (waypoint, stopover, destination) to define travel plans.
+void AI::IssueNPCOrders(Ship &ship, const System *targetSystem, const map<const Planet *, bool> stopovers)
+{
+	Orders newOrders;
+	const System *from = ship.GetSystem();
+	if(targetSystem)
+	{
+		DistanceMap distance(ship, targetSystem);
+		if(!distance.HasRoute(targetSystem))
+			ship.EraseWaypoint(targetSystem);
+		else
+		{
+			newOrders.type = Orders::TRAVEL_TO;
+			newOrders.targetSystem = targetSystem;
+			if(from == targetSystem)
+			{
+				// Travel to the next destination, if it exists.
+				ship.SetTargetStellar(nullptr);
+				const System *nextSystem = ship.NextWaypoint();
+				if(nextSystem)
+					newOrders.targetSystem = nextSystem;
+				else
+					newOrders.type = 0;
+			}
+		}
+	}
+
+	// If one of the planets in this system is a destination or stopover, it
+	// supercedes the order to travel to the next waypoint (unless already visited).
+	if(!stopovers.empty())
+		for(const auto &it : stopovers)
+			if(!it.second && it.first->IsInSystem(from))
+			{
+				newOrders.type = Orders::LAND_ON;
+				newOrders.targetPlanet = it.first;
+				break;
+			}
```

Comment:
Just noticed that this doesn't see any difference between a stopover and a destination, making it possible for an NPC to land on its supposed destination as if it was a stopover and then land forever on a supposed stopover.

I think an NPC should land on its destination only if it has visited every waypoint and stopover it had assigned.
Or alternatively, it could be able to land on the destination before completing other travel instructions, but that should result in the `land` objective being satisfied.

Reply:
Whoops, yes, it shouldn't be able to succeed until it's landed on all stopovers.